feat: add readme for AOD and template user guide (#6)
abcxyz/aod-template: README.md

# repo-template

**This is not an official Google product.**

> This template is for creating an AOD (access on demain) repository for GCP organizations, folders, and projects.

Reviewer: ```suggestion
This template is for creating an AOD (access on demand) repository to get access to GCP resources on demand.
```

> After completing the installation. Follow below steps to file an AOD request.
> 1. Open a PR
>   - A title and description to provide enough justification. E.g. "AOD request: debug customer issue X".
>   - Add following file at the repo root, note that any changes with below file changes will be considered AOD request and thus cannot be merged:
>     - iam.yaml - Request for IAM permissions on organization / folder / project level. To see an example of this file, refer to the example-iam.yaml file.
>   - Optionally use predefined duration labels on the PR to specify the IAM permission expiration. The duration is default to 2h if no label is used.
> 
> 2. PR approval
>   - Github workflows will check if the PR is an AOD request, marked by a "do_not_merge" failure workflow, and if it is a valid AOD request.
>   - Upon PR approval, the request will be handled by AOD.
> 
> 3. Close the PR after the request is handled successfully.

Reviewer: ```suggestion
## Make an AOD Request

**Make sure you have completed [prerequisite steps](link to the section below) to set up this repo properly.**

1. Open a PR
  - A title and description to provide enough justification. E.g. "AOD request: debug customer issue X".
  - To request IAM permissions on org/folder/project level, add an `iam.yaml` file in the repo root.
    - See an example of this file, see example in [example-iam.yaml](link to file).
    - Optionally use predefined duration labels on the PR to specify the IAM permission expiration. Otherwise a 2h default duration will be used.
  - (Later) To request on-demand `gcloud` commands, add an `gcloud.yaml` file in the repo root.

2. Checks
  - All AOD PRs will fail the check "do_not_merge" to prevent accidental merge.
  - The "Validate Request" check will report errors if there are problems in your `iam.yaml` or `gcloud.yaml`.
  
3. Getting approval
  - Ask one of the repo code owners to approve the PR
  - Upon approval, another GitHub workflow will run to grant you the IAM permissions or run the on-demand `gcloud` commands

3. Close the PR after you no longer need the access. The PR will automatically be closed after X hours (TODO).
```

> ## Installation

Reviewer: ```suggestion
## Prerequisites
```

> ## Installation

Reviewer: Add a section above about "How AOD works" and leave a TODO for me to fill in the content

> 

Reviewer: Define "who" should be carrying out these steps - likely someone with admin role on the project/folder/org

1.  Create an AOD repository using this template, only copy main branch is requried.
2.  Setup [Workload Identity Federation](https://cloud.google.com/iam/docs workload-identity-federation).
  - Create workload identity pool and provider in a GCP project where appropriate.
  - Create a GCP service account that identities from this pool can impersonate.
  - Grant the service account the required IAM roles to perform AOD operations.
    -  The minimum permissions are the get and set IAM policy. These permissions can be on projects, folders, or organizations level. For example, ["roles/resourcemanager.projectIamAdmin"](https://cloud.google.com/resource-manager/docs/access-control-proj#resourcemanager.projectIamAdmin) is a predefined role for managing project level IAM permission, an alternative is to use custom roles for a more restricted set of permissions.

Below is an example of how you setup Workload Identity Federation via terraform.

> ```terraform

Reviewer: Can we reference some of existing module instead of dumping all the Terraform here?

Reviewer: Terraform also isn't easy to understand. How about we delete these for now and follow up. I'm thinking referencing https://github.com/google-github-actions/auth#setting-up-workload-identity-federation first then reference terraform.

}

output "service_account_email" {
  description = "WIF service account identity email address."
  value       = google_service_account.wif_service_account.email
}
```

3.  Follow steps [here](https://docs.github.com/en/actions/learn-github-actions/variables#creating-configuration-variables-for-a-repository) to set the repository variables for WORKLOAD_IDENTITY_PROVIDER and SERVICE_ACCOUNT with outputs from step 2.

4. (Optional) Create labels to be used as IAM permission expiration. IAM permissions granted via AOD will have a default 2 hour expiration when the request is approved. To use custom expirations, follow steps [here](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#creating-a-label) to create duration labels. "duration-4h" is an example of valid duration label.

5. It is critical to enable the following repo settings:
  - Disable forking
  - Branch protection on main branch
  - Require approvals
  - Require review from Code Owners
  - Disallow specific actors to bypass required pull requests
  - Require status checks to pass before merging
  - Require signed commits
  - Disallow force pushes
  - Disallow deletions

>   - Set up CODEOWNERS with the group to approve AOD requests

Reviewer: Add a section below about adding code other than AOD. Could be a follow up to fill in the content

1. Open a PR
  - A title and description to provide enough justification. E.g. "AOD request: debug customer issue X".
  - To request IAM permissions on org/folder/project level, add an `iam.yaml` file in the repo root.
    - See an example of this file, see example in [example-iam.yaml](https://github.com/abcxyz/aod-template/example-iam.yaml).
    - Optionally use predefined duration labels on the PR to specify the IAM permission expiration. Otherwise a 2h default duration will be used.
  - (Later) To request on-demand `gcloud` commands, add an `gcloud.yaml` file in the repo root.

2. Checks
  - All AOD PRs will fail the check "do_not_merge" to prevent accidental merge.
  - The "Validate Request" check will report errors if there are problems in your `iam.yaml` or `gcloud.yaml`.

3. Getting approval
  - Ask one of the repo code owners to approve the PR
  - Upon approval, another GitHub workflow will run to grant you the IAM permissions or run the on-demand `gcloud` commands

4. Close the PR after you no longer need the access. The PR will automatically be closed after X hours (TODO).

## How AOD works(TODO)

> ## Prerequisites

Reviewer: Also add a TODO to revamp the prerequisites

Reviewer: Create some corresponding issues

# repo-template

**This is not an official Google product.**

This template is for creating an AOD (access on demand) repository to get access
to GCP resources on demand.

## Make an AOD Request

**Make sure you have completed
[prerequisite steps](https://github.com/abcxyz/aod-template/blob/main/README.md#prerequisites)
to set up this repo properly.**

1.  Open a PR

    -   A title and description to provide enough justification. E.g. "AOD
        request: debug customer issue X".
    -   To request IAM permissions on org/folder/project level, add an
        `iam.yaml` file in the repo root.
    -   See an example of this file, see example in

>         [example-iam.yaml](https://github.com/abcxyz/aod-template/example-iam.yaml).

Reviewer 2: The link showed 404.

>     -   Optionally use predefined duration labels on the PR to specify the IAM

Reviewer 2: (Optional)

        permission expiration. Otherwise a 2h default duration will be used.

>     -   (Later) To request on-demand `gcloud` commands, add an `gcloud.yaml`

Reviewer 2: Do we have an example of `gcloud.yaml`?

Author: Not really, it is to be implemented. The design doc has a rough example, but since this feature is not available yet, plan to add that when it id done.

        file in the repo root.

2.  Checks

    -   All AOD PRs will fail the check "do_not_merge" to prevent accidental
        merge.
    -   The "Validate Request" check will report errors if there are problems in
        your `iam.yaml` or `gcloud.yaml`.

3.  Getting approval

    -   Ask one of the repo code owners to approve the PR
    -   Upon approval, another GitHub workflow will run to grant you the IAM
        permissions or run the on-demand `gcloud` commands

4.  Close the PR after you no longer need the access. The PR will automatically
    be closed after X hours (TODO #8).

## How AOD works(TODO #7)

## Prerequisites

The admin of your GCP project/folder/org to complete the steps below.

1.  Create an AOD repository using this template, only copy main branch is
    required.

> 2.  Setup

Reviewer 2: Set up  [Workload Identity Federation](https://cloud.google.com/iam/docs/workload-identity-federation) by following guide [here](https://github.com/google-github-actions/auth#setting-up-workload-identity-federation)?

> 2.  Setup

Reviewer 2: Set up [Workload Identity Federation](https://cloud.google.com/iam/docs/workload-identity-federation) by following guide [here](https://github.com/google-github-actions/auth#setting-up-workload-identity-federation)?

Author: For now yes, we don't have a better guide for WIF yet.

Reviewer 2: I mean you can change the sentence to `Set up [Workload Identity Federation](https://cloud.google.com/iam/docs/workload-identity-federation) by following guide [here](https://github.com/google-github-actions/auth#setting-up-workload-identity-federation)`

    [Workload Identity Federation](https://cloud.google.com/iam/docs/workload-identity-federation),
    and a service account, see guide
    [here](https://github.com/google-github-actions/auth#setting-up-workload-identity-federation).

>     -   The minimum permissions are the get and set IAM policy. These

Reviewer 2: The minimum permissions are `getIamPolicy` and `setIamPolicy` and these permissions can be on projects, folders, or organizations level.

        permissions can be on projects, folders, or organizations level. For
        example,
        ["roles/resourcemanager.projectIamAdmin"](https://cloud.google.com/resource-manager/docs/access-control-proj#resourcemanager.projectIamAdmin)
        is a predefined role for managing project level IAM permission, an
        alternative is to use custom roles for a more restricted set of
        permissions.

3.  Follow steps
    [here](https://docs.github.com/en/actions/learn-github-actions/variables#creating-configuration-variables-for-a-repository)
    to set the repository variables for WORKLOAD_IDENTITY_PROVIDER and
    SERVICE_ACCOUNT with outputs from step 2.

4.  (Optional) Create labels to be used as IAM permission expiration. IAM
    permissions granted via AOD will have a default 2 hour expiration when the
    request is approved. To use custom expirations, follow steps
    [here](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#creating-a-label)

>     to create duration labels. "duration-4h" is an example of valid duration

Reviewer 2: Make `duration-4h` more obvious by changing the format like `duration-4h`

    label.

5.  It is critical to enable the following repo settings:

    -   Disable forking
    -   Branch protection on main branch
    -   Require approvals

>     -   Require review from Code Owners

Reviewer 2: Add Code Owners link?

    -   Disallow specific actors to bypass required pull requests
    -   Require status checks to pass before merging
    -   Require signed commits
    -   Disallow force pushes
    -   Disallow deletions

>     -   Set up CODEOWNERS with the group to approve AOD requests

Reviewer 2: add CODEOWNERS link? and put it before `Require review from Code Owners`
